mcs/class/System/Mono.AppleTls/AppleTlsContext.cs

```diff
@@ -0,0 +1,931 @@
+
+		public SslProtocol GetNegotiatedProtocolVersion ()
+		{
+			SslProtocol value;
+			var result = SSLGetNegotiatedProtocolVersion (Handle, out value);
+			CheckStatusAndThrow (result);
+			return value;
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern static /* OSStatus */ SslStatus SSLGetSessionOption (/* SSLContextRef */ IntPtr context, SslSessionOption option, out bool value);
+
+		public bool GetSessionOption (SslSessionOption option)
+		{
+			bool value;
+			var result = SSLGetSessionOption (Handle, option, out value);
+			CheckStatusAndThrow (result);
+			return value;
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern static /* OSStatus */ SslStatus SSLSetSessionOption (/* SSLContextRef */ IntPtr context, SslSessionOption option, bool value);
+
+		public void SetSessionOption (SslSessionOption option, bool value)
+		{
+			var result = SSLSetSessionOption (Handle, option, value);
+			CheckStatusAndThrow (result);
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern static /* OSStatus */ SslStatus SSLSetClientSideAuthenticate (/* SSLContextRef */ IntPtr context, SslAuthenticate auth);
+
+		public void SetClientSideAuthenticate (SslAuthenticate auth)
+		{
+			var result = SSLSetClientSideAuthenticate (Handle, auth);
+			CheckStatusAndThrow (result);
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern static /* OSStatus */ SslStatus SSLHandshake (/* SSLContextRef */ IntPtr context);
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern static /* OSStatus */ SslStatus SSLGetSessionState (/* SSLContextRef */ IntPtr context, ref SslSessionState state);
+
+		public SslSessionState SessionState {
+			get {
+				var value = SslSessionState.Invalid;
+				var result = SSLGetSessionState (Handle, ref value);
+				CheckStatusAndThrow (result);
+				return value;
+			}
+		}
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLGetPeerID (/* SSLContextRef */ IntPtr context, /* const void** */ out IntPtr peerID, /* size_t* */ out nint peerIDLen);
+
+		[DllImport ("/System/Library/Frameworks/Security.framework/Security")]
+		extern unsafe static /* OSStatus */ SslStatus SSLSetPeerID (/* SSLContextRef */ IntPtr context, /* const void* */ byte* peerID, /* size_t */ nint peerIDLen);
+
+		public unsafe byte[] PeerId {
+			get {
+				nint length;
+				IntPtr id;
+				var result = SSLGetPeerID (Handle, out id, out length);
+				CheckStatusAndThrow (result);
+				if ((result != SslStatus.Success) || ((int)length == 0))
+					return null;
+				var data = new byte [(int)length];
+				Marshal.Copy (id, data, 0, (int) length);
+				return data;
+			}
+			set {
+				SslStatus result;
+				nint length = (value == null) ? (IntPtr)0 : (IntPtr)value.Length;
```

Comment:
Call me old-fashioned, but I really don't like this - coming from a C background, I grew up distinguishing between pointers and numbers and I don't think turning `size_t` into a pointer is the right thing.

This is the native API that we're binding:
https://developer.apple.com/reference/security/1395882-sslgetpeerid?language=objc